### PR TITLE
adding bookmarking functionality 

### PR DESCRIPTION
**Pull request by anniehuo**
### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
- bookmarking jobcards on main feed screen is reflected in firebase and bookmarks screen in live time
- unbookmarking jobs on bookrmarks screen is also reflected in firebase and feed screen
- deleting job removes job from all user docs in firebase who had job bookmarked
- change bookmarks type added into authcontext that dispatches whenever a bookmark is toggled in job card, updates userobject bookmarkedjobs array that is a dependency of feed screen and bookmarks screen useeffect

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy!"
ezgif com-video-to-gif

## How to review
[//]: # "Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?"
- try bookmarking and unbookmarking jobs on feed screen and see if the correct jobcards show up on the bookmarks screen as well as the firebase
- make sure bookmarking icons properly reflect current bookmark/unbookmarked state of jobcards
- do everything above but do it very fast
- create a new job, bookmark it, then delete the job, check to make sure jobid is removed from users likedjobs array

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
- potential opportunity for refactoring/restructuring deleting jobs to not have to loop through all users to remove a job from the users bookmarkedjobs array
- there's a lag in jobcard rendering after switching from feed screen to bookmarks screen

### Online sources
[//]: # "Optional - copy links to any tutorials or documentation that was useful to you when working on this PR"

### Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @surajr10 @themiachou8

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -47,7 +47,9 @@
     "react-native-phone-input": "^1.3.4",
     "react-native-phone-number-input": "^2.1.0",
     "react-native-safe-area-context": "4.3.1",
+    "react-native-svg": "12.3.0",
     "react-native-screens": "^3.20.0",
+    "react-native-svg": "12.3.0",
```
Comment (surajr10) [inline package.json]: think you should be able to remove this duplicate package

File: src/assets/empty-heart.png
Comment (surajr10) [inline src/assets/empty-heart.png:1]: remove

File: src/assets/ex.png
Comment (surajr10) [inline src/assets/ex.png:1]: remove

File: src/assets/filled-heart.png
Comment (surajr10) [inline src/assets/filled-heart.png:1]: remove

File: src/components/JobCard/JobCard.tsx [hunk 2]
```diff
@@ -26,21 +37,37 @@ function JobCard({
...
   setFilteredJobs,
+  bookmarkedJobs,
+  setBookmarkedJobs,
 }: JobCardProps) {
+  const { userObject } = useContext(AuthContext);
+  const [bookmarkedValue, setBookmarked] = useState<boolean>();
+  const userObjectToString = JSON.stringify(userObject?.bookmarkedJobs);
+
+  useEffect(() => {
+    const getBookmarked = () => {
```
Comment (surajr10) [inline src/components/JobCard/JobCard.tsx]: can rename this  function to getBookmarkValue

File: src/components/JobCard/JobCard.tsx [hunk 3]
```diff
@@ -26,21 +37,37 @@ function JobCard({
   setPendingJobs,
   filteredJobs,
   setFilteredJobs,
+  bookmarkedJobs,
+  setBookmarkedJobs,
 }: JobCardProps) {
+  const { userObject } = useContext(AuthContext);
+  const [bookmarkedValue, setBookmarked] = useState<boolean>();
```
Comment (surajr10) [inline src/components/JobCard/JobCard.tsx]: can change these to bookmarkValue and setBookmarkValue

File: src/components/JobCard/JobCard.tsx [hunk 4]
```diff
@@ -26,21 +37,37 @@ function JobCard({
...
+  const userObjectToString = JSON.stringify(userObject?.bookmarkedJobs);
+
+  useEffect(() => {
+    const getBookmarked = () => {
+      const bookmarks = getBookmarks(job.id, userObject);
```
Comment (surajr10) [inline src/components/JobCard/JobCard.tsx]: I think you can remove the userObject as an argument to the getBookmarks function. instead i might say to get the userObject in the user.ts file and use that with the function

File: src/components/JobCard/JobCard.tsx [hunk 5]
```diff
@@ -26,21 +37,37 @@ function JobCard({
...
+      setBookmarked(bookmarks);
+    };
+    getBookmarked();
+  }, [job.id, userObjectToString, userObject]);
```
Comment (surajr10) [inline src/components/JobCard/JobCard.tsx]: not entirely sure which of these dependencies are required for this useEffect to work properly, but i might recommend to try taking them out 1 by 1 to see if things still work. this is for cleanliness and to ensure this doesn't get triggered when we don't want it to

File: src/components/JobCard/JobCard.tsx [hunk 6]
```diff
@@ -50,9 +77,28 @@ function JobCard({
...
       console.log(e);
     }
-    setFilteredJobs(filteredJobs.filter((_, index) => index !== idx));
+    if (filteredJobs === null) {
+      return;
+    }
+    setFilteredJobs?.(filteredJobs.filter((_, index) => index !== idx));
   }
 
+  const toggleBookmark = async (val: boolean | undefined) => {
```
Comment (surajr10) [inline src/components/JobCard/JobCard.tsx]: why is the type of val boolean or undefined

File: src/components/JobCard/JobCard.tsx [hunk 7]
```diff
@@ -50,9 +77,28 @@ function JobCard({
...
+    if (userObject !== null) {
+      if (userObject?.bookmarkedJobs?.includes(job.id)) {
```
Comment (surajr10) [inline src/components/JobCard/JobCard.tsx]: think this line can just be swapped with your getBookmarks function

File: src/components/JobCard/JobCard.tsx [hunk 8]
```diff
@@ -50,9 +77,28 @@ function JobCard({
...
+        const index = userObject?.bookmarkedJobs.indexOf(job.id);
```
Comment (surajr10) [inline src/components/JobCard/JobCard.tsx]: might recommend moving the logic on lines 89/90 to a helper function removeBookmarkedJob so it's easier to know what's being done here

File: src/firebase/firestore/user.ts [hunk 9]
```diff
@@ -46,12 +49,12 @@ export const addUser = async (user: User): Promise<void> => {
 
 export const updateUser = async (
```
Comment (surajr10) [inline src/firebase/firestore/user.ts]: don't think this function is being used anywhere so can delete

File: src/firebase/firestore/job.ts [hunk 10]
```diff
@@ -108,14 +111,33 @@ export const getAllJobs = async (collectionName: string): Promise<Job[]> => {
   }
 };
 
-// export const updateJob = async (job_id: string): Promise<void> => {};
+export const approveOrDeclineJob = async (
```
Comment (surajr10) [inline src/firebase/firestore/job.ts]: for the functions approveOrDeclineJob and deleteJob I would say to just have a single deleteJob that takes in a collection name and deletes from a particular collection. then make the rest of the logic of deleteJob (of removing from all users bookmarked collection) to another function

File: src/firebase/firestore/user.ts [hunk 11]
```diff
@@ -83,10 +86,73 @@ export const checkAndAddUser = async (
...
-      likedJobs: [], // switched to string of jobIds to match Firebase
+      bookmarkedJobs: [], // switched to string of jobIds to match Firebase
       name: 'test phone',
       phoneNumber: assignPhoneNumber,
       verified: true,
     });
   }
 };
+
+export const updateBookmarks = async (
```
Comment (surajr10) [inline src/firebase/firestore/user.ts]: don't this function is being used. can delete

File: src/navigation/BottomTabNavigators/AdminBottomTabNavigator.tsx [hunk 12]
```diff
@@ -72,6 +73,13 @@ function AdminStack(): ReactElement {
           tabBarIcon: LogoutVariantIcon,
         }}
       />
+      <Tab.Screen
+        name="Bookmarks"
+        component={BookmarksStackNavigator}
+        options={{
+          tabBarIcon: LogoutVariantIcon,
```
Comment (surajr10) [inline src/navigation/BottomTabNavigators/AdminBottomTabNavigator.tsx]: can you change the icon to be our bookmark one

File: src/navigation/BottomTabNavigators/UserBottomTabNavigator.tsx [hunk 13]
```diff
@@ -37,6 +38,13 @@ function UserStack(): ReactElement {
           tabBarIcon: HomeOutlineIcon,
         }}
       />
+      <Tab.Screen
+        name="Bookmarks"
+        component={BookmarksStackNavigator}
+        options={{
+          tabBarIcon: LogoutVariantIcon,
```
Comment (surajr10) [inline src/navigation/BottomTabNavigators/UserBottomTabNavigator.tsx:51]: change to bookmark icon

File: src/screens/Bookmarks/Bookmarks.tsx [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+import { AuthContext } from '../../context/AuthContext';
+
+function BookmarksScreen({
+  navigation,
+}: BookmarksStackScreenProps<'BookmarksScreen'>) {
+  const [bookmarkedList, setBookmarkedList] = useState([] as Job[]);
+  const { userObject } = useContext(AuthContext);
+  const userBookmarkedJobs = userObject?.bookmarkedJobs;
+
+  const isFocused = useIsFocused();
```
Comment (surajr10) [inline src/screens/Bookmarks/Bookmarks.tsx]: is the useIsFocused hook necessary here

File: src/screens/Bookmarks/Bookmarks.tsx [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+  useEffect(() => {
+    const fetchJobs = async () => {
```
Comment (surajr10) [inline src/screens/Bookmarks/Bookmarks.tsx]: call this function fetchBookmarkedJobs

File: src/screens/Bookmarks/Bookmarks.tsx [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+      const data = await getAllBookmarks(userObject);
```
Comment (surajr10) [inline src/screens/Bookmarks/Bookmarks.tsx]: can call this function getBookmarkedJobs

File: src/screens/Bookmarks/Bookmarks.tsx [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    };
+    fetchJobs();
+  }, [isFocused, userObject?.id, userObject]);
+
+  useEffect(() => {
+    const unsubscribe = navigation.addListener('blur', async () => {
+      if (userObject?.id === undefined) {
+        return;
+      }
+      await updateUserBookmarks(userBookmarkedJobs, userObject?.id);
```
Comment (surajr10) [inline src/screens/Bookmarks/Bookmarks.tsx:35]: i think if you follow what the comment i wrote about the getBookmarks function then i think you actually don't need any arguments to this function and can just use the userObject you get in the user.ts file

File: src/screens/Bookmarks/Bookmarks.tsx [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+  const isFocused = useIsFocused();
+
+  useEffect(() => {
+    const fetchJobs = async () => {
+      const data = await getAllBookmarks(userObject);
+      setBookmarkedList(data);
+    };
+    fetchJobs();
+  }, [isFocused, userObject?.id, userObject]);
```
Comment (surajr10) [inline src/screens/Bookmarks/Bookmarks.tsx]: same comment as before, check to see which of these dependencies are necessary

File: src/screens/Bookmarks/Bookmarks.tsx (hunk 14 above)
Comment (anniehuo) [reply src/screens/Bookmarks/Bookmarks.tsx]: yes

File: src/components/JobCard/JobCard.tsx (hunk 4 above)
Comment (anniehuo) [reply src/components/JobCard/JobCard.tsx]: usecontext can only be called within a react component/screen so i don't think i can get the userObject in the user.ts file

File: src/components/JobCard/JobCard.tsx (hunk 5 above)
Comment (anniehuo) [reply src/components/JobCard/JobCard.tsx]: if I remove job.id or userObject from the dependency array i get the following eslint error:

React Hook useEffect has a missing dependency: 'job.id'. Either include it or remove the dependency array.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)

but they both aren't technically needed

**Review by surajr10 — APPROVED** [review]
